File: impl-go/no_0739_daily_temperatures/no_0739_daily_temperatures.go
package no_0739_daily_temperatures

import (
	"container/list"
)

func dailyTemperatures(T []int) []int {
	res := make([]int, len(T))
	stack := list.New()
	for i, temperature := range T {
		for ele := stack.Front(); ele != nil; ele = stack.Front() {
			idx := ele.Value.(int)
			// 如果在栈中找到一个比当前温度低的，那么说明比那个温度稍高的就是当前的温度
			if T[idx] < temperature {
				res[idx] = i - idx
				stack.Remove(ele)
			} else {
				// 栈中的元素是单调递减的，如果 T[idx] >= temperature, 那么剩下的也会 >=.
				break
			}
		}
		stack.PushFront(i)
	}
	return res
}

File: impl-rust/src/no_0739_daily_temperatures.rs
struct Solution;

use std::collections::LinkedList;

impl Solution {
    pub fn daily_temperatures(t: Vec<i32>) -> Vec<i32> {
        let mut res = vec![0; t.len()];
        let mut stack = LinkedList::new();

        for (i, temperature) in t.iter().enumerate() {
            let mut front = stack.front();
            while let Some(idx) = front {
                // if t[idx] < temperature, then remove front.
                if let Some(should_remove_front) = t.get(*idx).map(|temp| temp < temperature) {
                    if should_remove_front {
                        res[*idx] = (i - *idx) as i32;
                        stack.pop_front();
                    } else {
                        break;
                    }
                }
                front = stack.front();
            }
            stack.push_front(i);
        }

        res
    }

    // 参考：https://leetcode-cn.com/problems/daily-temperatures/solution/rust-dan-diao-stack-by-jack_/
    pub fn daily_temperatures2(t: Vec<i32>) -> Vec<i32> {
        let mut res = vec![0; t.len()];
        // 把下标和温度值都存在栈中，所以这种方法比上面那种占用更多内存。
        let mut stack: LinkedList<(usize, i32)> = LinkedList::new();

        for (i, temperature) in t.iter().enumerate() {
            while let Some((idx, temp)) = stack.front() {
                // if t[idx] < temperature, then remove front.
                if temp < temperature {
                    res[*idx] = (i - idx) as i32;
                    stack.pop_front();
                } else {
                    break;
                }
            }
            stack.push_front((i, *temperature));
        }

        res
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_daily_temperatures() {
        let res = Solution::daily_temperatures(vec![73, 74, 75, 71, 69, 72, 76, 73]);
        assert_eq!(res, vec![1, 1, 4, 2, 1, 1, 0, 0]);
    }

    #[test]
    fn test_daily_temperatures2() {
        let res = Solution::daily_temperatures2(vec![73, 74, 75, 71, 69, 72, 76, 73]);
        assert_eq!(res, vec![1, 1, 4, 2, 1, 1, 0, 0]);
    }
}
